include/mfmg/dealii_operator.hpp

```diff
@@ -28,413 +31,161 @@
   using vector_type = VectorType;
-  using operator_type =
-      DealIIMatrixOperator<sparsity_pattern_type, matrix_type, vector_type>;
+  using operator_type = Operator<vector_type>;
 
-public:
-  DealIIMatrixOperator(
-      std::shared_ptr<matrix_type> matrix,
-      std::shared_ptr<sparsity_pattern_type> sparsity_pattern = nullptr)
-      : _sparsity_pattern(sparsity_pattern), _matrix(matrix)
-  {
-  }
+  DealIIMatrixOperator(std::shared_ptr<matrix_type> matrix,
+                       std::shared_ptr<sparsity_pattern_type> sparsity_pattern);
 
   virtual size_t m() const override final { return _matrix->m(); }
 
   virtual size_t n() const override final { return _matrix->n(); }
 
   size_t nnz() const { return _matrix->n_nonzero_elements(); }
 
-  virtual void apply(const vector_type &x, vector_type &y) const override final
-  {
-    _matrix->vmult(y, x);
-  }
-
-  std::shared_ptr<operator_type> transpose() const
-  {
-    ASSERT_THROW_NOT_IMPLEMENTED();
-
-    return nullptr;
-  }
+  virtual void apply(vector_type const &x, vector_type &y) const override final;
 
-  std::shared_ptr<operator_type> multiply(const operator_type &operator_b) const
-  {
-    ASSERT_THROW_NOT_IMPLEMENTED();
+  virtual std::shared_ptr<operator_type> transpose() const override final;
 
-    return nullptr;
-  }
+  virtual std::shared_ptr<operator_type>
+  multiply(operator_type const &operator_b) const override final;
 
   std::shared_ptr<matrix_type> get_matrix() const { return _matrix; }
 
   virtual std::shared_ptr<vector_type>
-  build_domain_vector() const override final
-  {
-    ASSERT_THROW_NOT_IMPLEMENTED();
-
-    return nullptr;
-  }
+  build_domain_vector() const override final;
 
-  virtual std::shared_ptr<vector_type> build_range_vector() const override final
-  {
-    ASSERT_THROW_NOT_IMPLEMENTED();
-
-    return nullptr;
-  }
+  virtual std::shared_ptr<vector_type>
+  build_range_vector() const override final;
 
 private:
-  // TODO: for some reason, order here is important. If _sparsity_pattern after
-  // _matrix, it results in some dealii subscriptor throws.
-  std::shared_ptr<sparsity_pattern_type>
-      _sparsity_pattern; // this may only be necessary for dealii::SparseMatrix,
-                         // which does not store the pattern. For Trilinos
-                         // matrices, this is not needed.
+  // The sparsity pattern needs to outlive the sparse matrix, so we declare it
+  // first. This is only necessary for dealii for deal.II sparse matrix.
+  // Trilinos saves the Epetra_Map inside the sparse matrix.
+  std::shared_ptr<sparsity_pattern_type> _sparsity_pattern;
+
   std::shared_ptr<matrix_type> _matrix;
 };
 
-template <class VectorType>
-class DealIIMatrixOperator<dealii::TrilinosWrappers::SparsityPattern,
-                           dealii::TrilinosWrappers::SparseMatrix, VectorType>
-    : public DealIIOperator<VectorType>
+template <typename VectorType>
+class TrilinosMatrixOperator : public Operator<VectorType>
```

Comment:
I propose either renaming `TrilinosMatrixOperator` to `DealIITrilinosMatrixOperator` or dropping `DealII` from `DealIIMatrixOperator` for consistency. They both live in some `dealii` namespace so that should be fine. Judging of how this patch renames `DealIISmootherOperator` to `SmootherOperator`, I am fine changing `DealIIMatrixOperator` to simply `MatrixOperator`.

Reply:
But then how do you call the matrix operator that derives from `Operator`? I have used `TrilinosMatrixOperator` and `DealIIMatrixOperator` because when Damien saw `DealIIMatrixOperator`, he assumed that we had sth called `TrilinosMatrixOperator`

Reply:
Yeah, while we could differentiate by namespaces, I like calling the concept of matrix operator `MatrixOperator`. The only minor issue I have right now is that `TrilinosMatrixOperator` does not indicate that it is deal.ii specific. I guess the namespace does that, but that's a bit of extra effort. I could live with leaving it as it is, but would prefer `DealIITrilinosMatrixOperator` and then probably `DealIIDirectOperator` with `DealIISmootherOperator`. In any case, this all is hidden from Hierarchy.